test/classes.coffee

```diff
@@ -1367,19 +1367,330 @@ test "nested classes with super", ->
+  eq Outer.Inner.name(), 'Inner'
+
+
+test "mixed constructors with inheritance and ES6 super", ->  
+  identity = (f) -> f
+
+  class TopClass
+    constructor: (arg) ->
+      @prop = 'top-' + arg
+
+  ```
+  class SuperClass extends TopClass {
+    constructor (arg) {
+      identity(super('super-' + arg));
+    }
+  }
+  ```
+  class SubClass extends SuperClass
+    constructor: ->
+      identity super 'sub'
+
+  ok (new SubClass).prop is 'top-super-sub'
+
+test "ES6 static class methods can be overriden", ->
+  class A
+    @name: -> 'A'
+
+  class B extends A
+    @name: -> 'B'
+
+  eq A.name(), 'A'
+  eq B.name(), 'B'
+
+# If creating static by direct assignment rather than ES6 static keyword
+test "ES6 Static methods should set `this` to undefined // ES6 ", ->
+  class A
+    @test: -> 
+      eq this, undefined
+
+# Ensure that our object prototypes work with ES6
+test "ES6 prototypes can be overriden", ->
+  class A
+    className: 'classA'
+
+  ```
+  class B {
+    test () {return "B";};
+  }
+  ```
+  b = new B
+  a = new A
+  eq a.className, 'classA'
+  eq b.test(), 'B'
+  Object.setPrototypeOf(b, a)
+  eq b.className, 'classA'
+  # This shouldn't throw, 
+  # as we only change inheritance not object construction
+  # This may be an issue with ES, rather than CS construction?
+  #eq b.test(), 'B'
+
+  class D extends B  
+  B::test = () -> 'D'
+  eq (new D).test(), 'D'
+
+
+test "ES6 conformance to extending non-classes", ->
+  A = (@title) ->
+    'Title: ' + @
+
+  class B extends A
+  b = new B('caffeinated') 
+  eq b.title, 'caffeinated'
+
+  # Check inheritance chain 
+  A::getTitle = () -> @title
+  eq b.getTitle(), 'caffeinated'
+
+  throwsC = """
+  C = {title: 'invalid'}
+  class D extends {}
+  """
+  # TODO: This should catch on compile and message should be "class can only extend classes and functions."
+  throws -> CoffeeScript.run throwsC, bare: yes
+
+# TODO: Either restrict valid statements in class bodies OR insert 'use strict'?
+test "Class function environment should be in `strict mode`, ie as if 'use strict' was in use", ->
+  class A
+    # this might be a meaningless test, since these are likely to be runtime errors and different
+    # for every browser.  Thoughts?
+    constructor: () ->
+      # Ivalid: prop reassignment
+      @state = {prop: [1], prop: {a: 'a'}}
+      # eval reassignment
+      @badEval = eval;
+
+  # Should throw, but doesn't
+  a = new A 
+
+# TODO: new.target needs support
+test "ES6 support for new.target (functions and constructors)", ->
+  throwsA = """
+  class A 
+    constructor: () ->
+      a = new.target.name 
+  """
+  throws -> CoffeeScript.compile throwsA, bare: yes
+
+test "only one method named constructor allowed", ->
+  throwsA = """
+  class A 
+    constructor: (@first) ->
+    constructor: (@last) ->
+  """
+  throws -> CoffeeScript.compile throwsA, bare: yes
+
+# TODO: We can't expect super to be enforced for all constructors.  If it's mandetory, 
```

Comment:
I agree, though we should remove the check for now since it prohibits valid constructs like returning from a derived constructor, and then add an improved compile-time check later.